### PR TITLE
Cross platform support for virtual shields

### PR DESCRIPTION
**Pull request by sidwarkd**

### CODE REVIEW

**Comment by zfields** [comment]
@JimGaleForce Please code review this.

File: src/Graphics.cpp [hunk 1]
```diff
@@ -56,14 +55,14 @@ const PROGMEM char CHANGEACTION[] = "CHANGE";
 Graphics::Graphics(const VirtualShield &shield) : Text(shield) {
 }
 
-int Graphics::line(UINT x1, UINT y1, UINT x2, UINT y2, ARGB argb, UINT weight)
+int Graphics::line(unsigned int x1, unsigned int y1, unsigned int x2, unsigned int y2, ARGB argb, unsigned int weight)
```
Comment (ooeygui) [inline src/Graphics.cpp]: why not typedef these?

File: src/Graphics.cpp [hunk 2]
```diff
@@ -56,14 +55,14 @@ const PROGMEM char CHANGEACTION[] = "CHANGE";
...
 {
 	EPtr eptrs[] = { EPtr(ACTION, LINE), EPtr(Y, (uint32_t)y1), EPtr(X, (uint32_t)x1),
 		EPtr(X2, (uint32_t)x2), EPtr(Y2, (uint32_t)y2),
 		EPtr(RGBAKEY, (uint32_t)argb.color, argb.color ? Uint : None),
 		EPtr(WIDTH, (uint32_t) weight, weight == 1 ? None : Uint) };
 
-	return shield.block(writeAll(SERVICE_NAME_GRAPHICS, eptrs, 7), onEvent == 0);
+	return shield.block(writeAll(SERVICE_NAME_GRAPHICS, eptrs, 7), onEvent == NULL);
```
Comment (ooeygui) [inline src/Graphics.cpp]: how about nullptr? (can be #define on arduino)

File: src/VirtualShield.cpp [hunk 3]
```diff
@@ -420,12 +419,18 @@ void VirtualShield::onJsonStringReceived(char* json, ShieldEvent* shieldEvent) {
 /// <param name="length">The length.</param>
 /// <param name="shieldEvent">The shield event.</param>
 void VirtualShield::onStringReceived(char* buffer, int length, ShieldEvent* shieldEvent) {
+#if defined(_WINDOWS_)
+	char *json = new char[length + 1];
```
Comment (ooeygui) [inline src/VirtualShield.cpp]: For temporary local buffers, alloca might be a better option. It avoids thrashing the heap. Works on Arduino as well, so you don't need two happy paths

File: src/VirtualShield.cpp [hunk 4]
```diff
@@ -749,8 +757,9 @@ int VirtualShield::parseToHash(const char* text, unsigned int *hash, int hashCou
 // per Paul Larson - Microsoft Research
 unsigned int VirtualShield::hash(const char* s, unsigned int len, unsigned int seed)
 {
+    if (!s) { return seed; }
 	unsigned hash = seed;
-	while ((len == -1) ? *s : len-- > 0)
+	while ((len == static_cast<unsigned int>(-1)) ? *s : len-- > 0)
```
Comment (ooeygui) [inline src/VirtualShield.cpp]: I'd love to see this sentinel pulled out as a constant.

**Comment by ooeygui** [comment]
Please verify:
* Works on Arduino
* Works on Raspberry Pi
* Plans exist to update Readme.

**Comment by JimGaleForce** [comment]
logically looks good. question: how do users use their Arduino Strings? does this approach preserve the String usage (from an INO file) that we all discussed? thx

File: src/Orientation.h [hunk 5]
```diff
@@ -25,18 +25,18 @@
...
 
-#include "Arduino.h"
+#include "Sensor.h"
+#include "ShieldEvent.h"
+#include "VirtualShield.h"
 
 class Orientation : public Sensor {
 public:
-	double X;
+    double W;
```
Comment (ooeygui) [inline src/Orientation.h]: tabs?

File: src/Graphics.cpp (hunk 1 above)
Comment (ooeygui) [inline src/Graphics.cpp]: actually with stdint, you should use uint32_t.

File: src/VirtualShield.cpp [hunk 6]
```diff
@@ -782,13 +791,19 @@ int VirtualShield::directToSerial(const char* cmd)
...
 int VirtualShield::sendFlashStringOnSerial(const char* flashStringAdr, int start, bool encode) const
 {
-	unsigned char dataChar = 0;
-	const int actualStart = start < 0 ? 0 : start;
+	char dataChar = '\0';
+	const size_t actualStart = start < 0 ? 0 : start;
 	const bool isFormatted = start > DEFAULT_LENGTH;
 
-	for (int i = actualStart; i < strlen_PF((uint_farptr_t)flashStringAdr); i++)
+#if defined(_WINDOWS_)
```
Comment (ooeygui) [inline src/VirtualShield.cpp]: Add the far pointer PF support to pgmspace.h in lightning. https://github.com/ms-iot/lightning/blob/develop/source/avr/pgmspace.h
then remove this ifdef. It should just work in lightning.

**Comment by ooeygui** [comment]
:+1: 
Next change: 
* Fix far progmem in lightning. 
* Look into alloca or single static large heap block.

**Pull request merged**